core/arch/atomic_exports.h

```diff
@@ -723,7 +726,296 @@ atomic_dec_becomes_zero(volatile int *var)
+                                     : "memory");                   \
+            } while (0)
+#        define ATOMIC_1BYTE_WRITE(target, value, hot_patch)   \
+            do {                                               \
+                /* Not currently used to write code */         \
+                ASSERT_CURIOSITY(!hot_patch);                  \
+                /* Fence for release semantics. */             \
+                __asm__ __volatile__("fence iorw,iorw\n\t"     \
+                                     "sb %0, 0(%1)\n\t"        \
+                                     :                         \
+                                     : "r"(value), "r"(target) \
+                                     : "memory");              \
+            } while (0)
+#        define ATOMIC_4BYTE_ALIGNED_WRITE(target, value, hot_patch) \
+            do {                                                     \
+                /* Page 25 of Volume I: RISC-V Unprivileged ISA      \
+                 * V20191213 states that aligned loads/stores are    \
+                 * atomic. Alternatively we could use                \
+                 * amoswap.w zero %0, 0(%1) but that would incur     \
+                 * an extra read we don't need and also requires     \
+                 * alignment.                                        \
+                 */                                                  \
+                ASSERT(ALIGNED(target, 4));                          \
+                /* Not currently used to write code. */              \
+                ASSERT_CURIOSITY(!hot_patch);                        \
+                /* Fence for release semantics. */                   \
+                __asm__ __volatile__("fence iorw,iorw\n\t"           \
+                                     "sw %0, 0(%1)\n\t"              \
+                                     :                               \
+                                     : "r"(value), "r"(target)       \
+                                     : "memory");                    \
+            } while (0)
+#        define ATOMIC_4BYTE_WRITE ATOMIC_4BYTE_ALIGNED_WRITE
+#        define ATOMIC_4BYTE_ALIGNED_READ(addr_src, addr_res)       \
+            do {                                                    \
+                /* Page 25 of Volume I: RISC-V Unprivileged ISA     \
+                 * V20191213 states that aligned loads/stores are   \
+                 * atomic.                                          \
+                 */                                                 \
+                ASSERT(ALIGNED(addr_src, 4));                       \
+                ASSERT(ALIGNED(addr_res, 4));                       \
+                /* Fence for acquire semantics. */                  \
+                __asm__ __volatile__("lw %0, 0(%1)\n\t"             \
+                                     "fence iorw,iorw"              \
```

Comment:
If I do:
```C
static std::atomic<int> myatomic;
...
    if (myatomic.load(std::memory_order_acquire) == 0) {
```
It seems to compile to:
```
 98c:   0ff0000f                fence
 990:   439c                    lw      a5,0(a5)
 992:   0ff0000f                fence
```
With a `fence` before as well as after the load?

Update: that's -O0; compiling -O1 removes the first fence.  Hmm.  Never mind.